### PR TITLE
Started implementation of hunter creation with test setup

### PR DESCRIPTION
**Pull request by Davidde94**

### CODE REVIEW

File: Tests/SpartaLibTests/HunterTests.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
+//
+//  HunterTests.swift
+//  SpartaTests
+//
+//  Created by David Evans on 21/07/2018.
+//
```
Comment (RyanMKrol) [inline Tests/SpartaLibTests/HunterTests.swift]: Add a song

File: Sources/SpartaLib/Users/User.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    var surname: String { get set }
+    var email: String { get }
+    var auth: String { get }
+    
+    func updateEmail(_ email: String) throws
+    func updatePassword(_ password: String) throws
+    
+}
+
+extension User {
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/User.swift]: why is this an extension

File: Sources/SpartaLib/Users/Hunter.swift [hunk 3]
```diff
@@ -0,0 +1,64 @@
+//
+//  Hunter.swift
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/Hunter.swift]: Not a fan of hunters

File: Package.swift [hunk 4]
```diff
@@ -10,10 +10,13 @@ let package = Package(
     ],
     targets: [
         .target(
-            name: "Sparta",
+            name: "SpartaLib",
```
Comment (RyanMKrol) [inline Package.swift]: why is there a sparta lib?

File: Sources/SpartaLib/Users/User.swift [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+//  Sparta
+//
+//  Created by David Evans on 21/07/2018.
+//
+
+// Song: None (but watching Top Gear)
+
+import Foundation
+
+protocol User: AnyObject {
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/User.swift]: Why do you need to make this an AnyObject

File: Tests/SpartaLibTests/HunterTests.swift [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+@testable import SpartaLib
+
+class HunterTests: XCTestCase {
+
+    func testCreationFromInfoStruct() {
+        
+        let bundle = Bundle(for: type(of: self))
+        print(bundle.bundlePath)
+        let dataPath = bundle.path(forResource: "testHunter", ofType: "json")!
+        let data = try! Data(contentsOf: URL(fileURLWithPath: dataPath))
```
Comment (RyanMKrol) [inline Tests/SpartaLibTests/HunterTests.swift]: i think we should just be creating a struct, and passing it in, for our unit tests. This is overkill. If you want to check that the encoding works, that's a different test.

File: Sources/SpartaLib/Users/Hunter.swift [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ request without extra code while allowing for extra
+ pieces of information that shouldnt be stored inside
+ Hunter itself (e.g. the password).
+ */
+struct HunterOnboardingInfo: Codable {
+    let forename: String
+    let surname: String
+    let email: String
+    let auth: String
+    let password: String
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/Hunter.swift]: I think some of these are optional if you choose facebook login type right?

File: Tests/SpartaLibTests/HunterTests.swift (hunk 1 above)
Comment (Davidde94) [reply Tests/SpartaLibTests/HunterTests.swift]: Oh come on

File: Sources/SpartaLib/Users/User.swift (hunk 2 above)
Comment (Davidde94) [reply Sources/SpartaLib/Users/User.swift]: You can't have computer variables inside a protocol definition, has to be an extension.

File: Sources/SpartaLib/Users/Hunter.swift (hunk 3 above)
Comment (Davidde94) [reply Sources/SpartaLib/Users/Hunter.swift]: Not a fan of you

File: Package.swift (hunk 4 above)
Comment (Davidde94) [reply Package.swift]: To unit test. You can't unit test an executable, so this is the method of doing it in Swift. The executable contains nothing other than `main.swift`, and the library (which is kept private to this repo) contains all the good stuff, which can then be unit tested.

File: Tests/SpartaLibTests/HunterTests.swift (hunk 1 above)
Comment (RyanMKrol) [reply Tests/SpartaLibTests/HunterTests.swift]: Every. File.

File: Sources/SpartaLib/Users/User.swift (hunk 5 above)
Comment (Davidde94) [reply Sources/SpartaLib/Users/User.swift]: Safer, we never want a user to be anything other than an object (e.g. without this `AnyObject constraint) you could theoretically create an enum that is a user, which would be weird.

File: Sources/SpartaLib/Users/Hunter.swift (hunk 3 above)
Comment (RyanMKrol) [reply Sources/SpartaLib/Users/Hunter.swift]: Fair, it's growing on me actually. Let's call agents Vultures.

File: Tests/SpartaLibTests/HunterTests.swift (hunk 6 above)
Comment (Davidde94) [reply Tests/SpartaLibTests/HunterTests.swift]: This was more of a test setup test, but I'll update.

File: Sources/SpartaLib/Users/User.swift (hunk 5 above)
Comment (RyanMKrol) [reply Sources/SpartaLib/Users/User.swift]: Huh, fair play

File: Sources/SpartaLib/Users/Hunter.swift (hunk 7 above)
Comment (Davidde94) [reply Sources/SpartaLib/Users/Hunter.swift]: Nope, all still required. The password is the Facebook authentication token.

File: Sources/SpartaLib/Users/User.swift (hunk 2 above)
Comment (RyanMKrol) [reply Sources/SpartaLib/Users/User.swift]: Ahh nice

File: Package.swift (hunk 4 above)
Comment (RyanMKrol) [reply Package.swift]: Fair

File: Sources/SpartaLib/Users/Hunter.swift (hunk 7 above)
Comment (RyanMKrol) [reply Sources/SpartaLib/Users/Hunter.swift]: Hmmm, maybe these should be separate things then, it's not immediately obvious that the password could be a password or a token

File: Tests/SpartaLibTests/HunterTests.swift (hunk 6 above)
Comment (RyanMKrol) [reply Tests/SpartaLibTests/HunterTests.swift]: Damn right you will

File: Sources/SpartaLib/Users/Hunter/HunterOnboarding.swift [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    var surname: String { get }
+    var email: EmailType { get }
+    var auth: String { get }
+    var password: String { get }
+    var loginType: HunterLoginType { get }
+}
+
+struct HunterEmailOnboardingInfo: HunterOnBoardingInfo {
+    
+    typealias EmailType = String
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/Hunter/HunterOnboarding.swift]: Do we need a type alias here? seems like overkill

File: Sources/SpartaLib/Users/Hunter/HunterOnboarding.swift [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+struct HunterFacebookOnboardingInfo: HunterOnBoardingInfo {
+    
+    typealias EmailType = String?
+    
+    let forename: String
+    let surname: String
+    let email: EmailType
+    let auth: String
+    let password: String
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/Hunter/HunterOnboarding.swift]: Why does this still have a password?

File: Sources/SpartaLib/Users/Hunter/Hunter.swift [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+class Hunter: User, Codable {
+    
+    var forename: String
+    var surname: String
+    private(set) var email: String?
+    private(set) var auth: String
+    let loginType: HunterLoginType
+    
+    init<T: HunterOnBoardingInfo>(_ onboarding: T) throws {
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/Hunter/Hunter.swift]: Are we not doing anything with password here?

File: Sources/SpartaLib/Users/Hunter/Hunter.swift [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        email = onboarding.email as? String
+        auth = onboarding.auth
+        loginType = onboarding.loginType
+    }
+    
+    func updateEmail(_ email: String) throws {
+        
+    }
+    
+    func updatePassword(_ password: String) throws {
```
Comment (RyanMKrol) [inline Sources/SpartaLib/Users/Hunter/Hunter.swift]: Should facebook users be able to update password?

File: Tests/SpartaLibTests/TestHelpers.swift [hunk 12]
```diff
@@ -0,0 +1,23 @@
+//
+//  TestHelpers.swift
+//  SpartaLib
+//
+//  Created by David Evans on 22/07/2018.
+//
```
Comment (RyanMKrol) [inline Tests/SpartaLibTests/TestHelpers.swift]: Song

**Pull request merged**